### ISSUE TITLE
ThreeES6 - running XMLHTTPRequests after dispose()

### ISSUE THREAD

**Issue by jfeick** [cnr-isti-vclab/nexus#106]
First of all, thank you very much for your continuous efforts to provide a three.js ES6 implementation :)

I noticed that a call to `dipose()` cleans up the cache node tree but does not abort running XMLHTTPRequests that have not already been handled as nodes. This manifests in weird errors if you are using a texture as scene background in three.js as it overwrites the texture-buffer with one from the nexus node. Furthermore, these non-aborted requests fail to decrease the `Cache.pending` member and can lead to a state where `Cache.pending >= maxPending` and thus blocking new network requests.

I tried to fix this behavior by extending the `flush(mesh)` method to iterate over `mesh.texreq` and `mesh.georeq` and abort danglings requests:

```javascript
    flush (mesh) {
        if (!this.nodes.has(mesh)) { return }
        for (const id of this.nodes.get(mesh))
            {
              // console.log('removeNode with id', id)
              this.removeNode(mesh, id)
            }
        for (const [id, req] of Object.entries(mesh.texreq)) {
          if (!req.readyState != 4) {
            req.abort()
            delete mesh.texreq[id]
          }
        }
        for (const [id, req] of Object.entries(mesh.georeq)) {
          if (!req.readyState != 4) {
            req.abort()
            delete mesh.georeq[id]
          }
        }
        this.nodes.delete(mesh)
    },
```

and I changed the abort callback in `requestNodeGeometry` and `requestNodeTexture` as follows:

```javascript
            function () {
                delete mesh.georeq[id]
                t.pending--
                // t.removeNode(mesh, id)
            },
```

The call to `removeNode` on an aborted node would result in exceptions, but I assume I forgot to clean up some buffer allocations that I am not aware of.

**Comment by jfeick** [comment]
Oh... I tested my "fix" with a faster connection, and the code to call abort() on the pending requests still does not solve the race condition. 

![textureleak](https://user-images.githubusercontent.com/2508010/113709358-b0ec8d00-96e2-11eb-9315-185b0e74d651.png)

**Comment by ponchio** [comment]
I had a look at the problem and the mistake was probably not to add nodes to the cache as soon as the node was requested instead of waiting for the data to arrive. In this way removeNode takes care of the problem.

Hopefully this fixes the issue. In case it doesn't, would you share your sample so I can replicate the issue?

**Comment by jfeick** [comment]
Thank you very much for your help, but there still seems to be an issue regarding the state of the `Cache`. 
For one, `deleteTexture(tex)` of `Nexus3D` throws, which I just commented out.
But the `Cache.pending` doesn't get updated properly after several loads/disposes of a Nexus model ending up with a state where `pending >= maxPending`. This results in a state where no new requests are made for a fresh model. 

I tried to add code to `removeNode(mesh, id)` to adjust the pending state, but with this I can still end up with negative values for pending:

```javascript
    removeNode (mesh, id) {
        this.nodes.get(mesh).delete(id)

        if (mesh.status[id] == 0) {
          if (this.debug.verbose)
          { console.log('Double remove due to abort.') }
          if (this.pending > 0) { this.pending-- }
          return // TODO
        }
     /* [...] */
```
Any idea where I'm going wrong? The problem is not specific to one Nexus-Model (which I can gladly provide), but does appear in an application where several Nexus models are loaded/unloaded by the user in a viewer application.
